### PR TITLE
Feat(Tree): indeterminate state

### PR DESCRIPTION
**Pull request by assuncaocharles**
#### Pull request checklist

- [ ] Addresses an existing issue: Fixes #0000
- [ ] Include a change request file using `$ yarn change`

#### Description of changes

**It depends on [feat(Tree): adding multiselection](https://github.com/microsoft/fluentui/pull/12362)**

It adds a 3rd state for checkbox in the tree, identifying when the groups has been partially selected. 

Screen Recording 2020-04-16 at 2 43 13 PM

#### Focus areas to test

(optional)

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/microsoft/fluentui/pull/12739)

### CODE REVIEW

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorUrl.ts [hunk 1]
```diff
@@ -4,3 +4,10 @@ export default (color: string, backgroundColor: string) =>
   )}; padding: 2px;' focusable='false' viewBox='8 8 22.5 22.5'%3E%3Cg%3E%3Cpath fill='${escape(
     color,
   )}' d='M23.5 11.875a.968.968 0 0 1-.289.711l-8.25 8.25c-.192.193-.43.289-.711.289s-.519-.096-.711-.289l-4.75-4.75a.965.965 0 0 1-.289-.711c0-.125.027-.25.082-.375s.129-.234.223-.328a.953.953 0 0 1 .695-.297c.135 0 .266.025.391.074.125.05.231.121.32.215l4.039 4.047 7.539-7.547a.886.886 0 0 1 .32-.215c.125-.049.255-.074.391-.074a1.004 1.004 0 0 1 .922.625.97.97 0 0 1 .078.375z' /%3E%3C/g%3E%3C/svg%3E")`;
+
+export const checkboxIndicatorIndeterminatedUrl = (color: string, backgroundColor: string) =>
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorUrl.ts:8]: Extract this to separate file and rename the initial one if needed

File: packages/fluentui/react-northstar/src/components/Tree/TreeTitle.tsx [hunk 2]
```diff
@@ -126,17 +186,27 @@ const TreeTitle: React.FC<WithAsProp<TreeTitleProps>> & FluentComponentStaticPro
 TreeTitle.deprecated_className = 'ui-tree__title';
 TreeTitle.displayName = 'TreeTitle';
 
+TreeTitle.slotClassNames = {
+  indicator: `${TreeTitle.deprecated_className}__selection-indicator`,
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Tree/TreeTitle.tsx:190]: is this used anywhere?

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorUrl.ts (hunk 1 above)
Comment (mnajdova) [reply packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorUrl.ts:8]: It's a bit confusing why it is added here, when it is used in the tree* components.

File: packages/fluentui/react-northstar/src/themes/teams/components/Tree/treeTitleVariables.ts [hunk 3]
```diff
@@ -1,13 +1,50 @@
...
+    borderColor: siteVars.colorScheme.default.foreground1,
+    borderStyle: 'solid',
+    borderRadius: pxToRem(3),
+    borderWidth: pxToRem(1),
+    indicatorColor: 'transparent',
+    selectionIndicatorMargin: `0 0 0 0`,
+    background: 'transparent',
+
+    checkedBackground: siteVars.colorScheme.brand.backgroundActive1,
+    checkedBorderColor: _.get(siteVars, 'colorScheme.brand.backgroundActive1', defaultValue),
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/themes/teams/components/Tree/treeTitleVariables.ts:42]:
```suggestion
    checkedBorderColor: siteVarscolorScheme.brand.backgroundActive1,
```

Same for the others variables

File: packages/fluentui/react-northstar/src/themes/teams/components/Tree/treeTitleVariables.ts [hunk 4]
```diff
@@ -1,13 +1,50 @@
...
+  checkedBorderColor: string;
+  checkedIndicatorColor: string;
+
+  disabledBackground: string;
+  disabledBackgroundChecked: string;
+  disabledBorderColor: string;
+  disabledCheckedIndicatorColor: string;
 }
 
+const defaultValue = 'red';
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/themes/teams/components/Tree/treeTitleVariables.ts:26]: remove this

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorIndeterminatedUrl.ts [hunk 5]
```diff
@@ -0,0 +1,6 @@
+export default (color: string, backgroundColor: string) =>
+  `url("data:image/svg+xml, %3Csvg xmlns='http://www.w3.org/2000/svg' role='presentation' style='background-color: ${escape(
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorIndeterminatedUrl.ts:2]: Move this to the component where it is used

File: packages/fluentui/react-northstar/src/components/Tree/TreeTitle.tsx [hunk 6]
```diff
@@ -72,15 +72,21 @@ export interface TreeTitleProps extends UIComponentProps, ChildrenComponentProps
 
   /** Whether or not tree title is selectable. */
   selectable?: boolean;
+
+  /** For selectable parents define if all nested children are checked */
+  incompleteChecked?: boolean;
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Tree/TreeTitle.tsx:77]: Why did you decide not to use `indeterminate`?

Comment (assuncaocharles) [reply packages/fluentui/react-northstar/src/components/Tree/TreeTitle.tsx:77]: Changed

File: packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorIndeterminatedUrl.ts (hunk 5 above)
Comment (assuncaocharles) [reply packages/fluentui/react-northstar/src/themes/teams/components/Checkbox/checkboxIndicatorIndeterminatedUrl.ts:2]: Done

**Comment by assuncaocharles** [comment]
> I see when selectable parent is collapsed, then we are displaying the checkbox
> it looks like regreession
> image

Fixed

**Comment by kolaps33** [comment]
When click on the "selectable parent" next to "House Targaryen", the all child was selected. 
Then I unselect these children. 

But parent didn't change to "unselected" state, but stay in indetrminate state. (Interesting the same steps with 'House Lannister' was working )

image

**Comment by assuncaocharles** [comment]
> When click on the "selectable parent" next to "House Targaryen", the all child was selected.
> Then I unselect these children.
> 
> But parent didn't change to "unselected" state, but stay in indetrminate state. (Interesting the same steps with 'House Lannister' was working )
> 
> image

Good catch. :)

Fixed

File: packages/fluentui/docs/src/examples/components/Toolbar/Visual/ToolbarExampleVariables.shorthand.tsx [hunk 7]
```diff
@@ -67,6 +66,7 @@ const ToolbarExampleVariables = () => {
           icon: <MoreIcon />,
           key: 'item-menu',
           menu: {
+            accessibility: notAutoFocusToolbarMenuBehavior,
```
Comment (kolaps33) [inline packages/fluentui/docs/src/examples/components/Toolbar/Visual/ToolbarExampleVariables.shorthand.tsx:69]: These changes are related to the issue
https://dev.azure.com/microsoftdesign/fluent-ui/_workitems/edit/7148

Comment (assuncaocharles) [reply packages/fluentui/docs/src/examples/components/Toolbar/Visual/ToolbarExampleVariables.shorthand.tsx:69]: Thanks

File: packages/fluentui/react-northstar/src/components/Tree/Tree.tsx [hunk 8]
```diff
@@ -350,6 +352,32 @@ class Tree extends AutoControlledComponent<WithAsProp<TreeProps>, TreeState> {
     });
   };
 
+  getAllSelectableChildrenId = items => {
+    return items.reduce((acc, item) => {
+      if (item.items) {
+        return [...acc, ...this.getAllSelectableChildrenId(item.items)];
+      }
+      return item.hasOwnProperty('selectable') && !item.item ? acc : [...acc, item.id];
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Tree/Tree.tsx:360]:
```suggestion
      return item.hasOwnProperty('selectable') && !item.items ? acc : [...acc, item.id];
```
Shouldn't this be `item.items`

File: packages/fluentui/react-northstar/src/components/Tree/Tree.tsx [hunk 9]
```diff
@@ -350,6 +352,32 @@ class Tree extends AutoControlledComponent<WithAsProp<TreeProps>, TreeState> {
...
+    }, []);
+  };
+
+  isIncompletedGroupChecked = (item: TreeItemProps) => {
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Tree/Tree.tsx:364]:
```suggestion
  isIndeterminate = (item: TreeItemProps) => {
```

File: packages/fluentui/react-northstar/src/components/Tree/Tree.tsx [hunk 10]
```diff
@@ -367,7 +395,8 @@ class Tree extends AutoControlledComponent<WithAsProp<TreeProps>, TreeState> {
         const { id } = item;
         const isSubtree = hasSubtree(item);
         const isSubtreeExpanded = isSubtree && this.isActiveItem(id);
-        const isSelectedItem = this.isSelectedItem(id);
+        const isSelectedItem = this.isSelectedItem(item);
+        const indeterminate = this.isIncompletedGroupChecked(item);
```
Comment (mnajdova) [inline packages/fluentui/react-northstar/src/components/Tree/Tree.tsx:399]:
```suggestion
        const indeterminate = this.isIndeterminate(item);
```

File: packages/fluentui/react-northstar/src/components/Tree/Tree.tsx (hunk 8 above)
Comment (assuncaocharles) [reply packages/fluentui/react-northstar/src/components/Tree/Tree.tsx:360]: The correct should be `item.hasOwnProperty('selectable') && !item.selectable`

File: packages/fluentui/react-northstar/src/components/Tree/Tree.tsx (hunk 10 above)
Comment (assuncaocharles) [reply packages/fluentui/react-northstar/src/components/Tree/Tree.tsx:399]: Done

**Pull request merged**